File: backend/src/controllers/watchlist.ts
/**
 * backend/src/controllers/watchlist.ts
 * HTTP Controller for watchlist configurations
 */

import type { Response, NextFunction } from 'express'
import type { AuthenticatedRequest } from '../middlewares/auth.js'
import { prisma } from '../db/prisma.js'

export async function getWatchlists(req: AuthenticatedRequest, res: Response, next: NextFunction): Promise<any> {
  try {
    const userId = req.user!.id
    const watchlists = await prisma.watchlist.findMany({
      where: { userId },
      include: { items: true },
      orderBy: { createdAt: 'desc' }
    })
    
    return res.status(200).json({ success: true, watchlists })
  } catch (error) {
    next(error)
  }
}

export async function createWatchlist(req: AuthenticatedRequest, res: Response, next: NextFunction): Promise<any> {
  try {
    const userId = req.user!.id
    const { name } = req.body

    if (!name || !name.trim()) {
      return res.status(400).json({ error: true, message: 'Please provide a valid watchlist name.' })
    }

    const newWl = await prisma.watchlist.create({
      data: {
        userId,
        name: name.trim()
      },
      include: { items: true }
    })

    return res.status(201).json({ success: true, watchlist: newWl })
  } catch (error) {
    next(error)
  }
}

export async function addItem(req: AuthenticatedRequest, res: Response, next: NextFunction): Promise<any> {
  try {
    const userId = req.user!.id
    const watchlistId = req.params.id
    const { symbol, targetPrice, alertEnabled } = req.body

    if (!symbol) {
      return res.status(400).json({ error: true, message: 'Please specify a company stock symbol.' })
    }

    // Verify watchlist ownership
    const wl = await prisma.watchlist.findFirst({
      where: { id: watchlistId, userId }
    })

    if (!wl) {
      return res.status(404).json({ error: true, message: 'Watchlist not found or unauthorized.' })
    }

    // Check if item already exists in watchlist
    const existing = await prisma.watchlistItem.findFirst({
      where: { watchlistId, symbol: symbol.toUpperCase() }
    })

    if (existing) {
      return res.status(409).json({ error: true, message: 'Stock already exists inside this watchlist.' })
    }

    const newItem = await prisma.watchlistItem.create({
      data: {
        watchlistId,
        symbol: symbol.toUpperCase(),
        targetPrice: targetPrice ? parseFloat(targetPrice) : null,
        alertEnabled: alertEnabled ?? false
      }
    })

    return res.status(201).json({ success: true, item: newItem })
  } catch (error) {
    next(error)
  }
}

export async function updateItem(req: AuthenticatedRequest, res: Response, next: NextFunction): Promise<any> {
  try {
    const userId = req.user!.id
    const { itemId } = req.params
    const { targetPrice, alertEnabled } = req.body

    // Verify ownership
    const item = await prisma.watchlistItem.findFirst({
      where: {
        id: itemId,
        watchlist: { userId }
      }
    })

    if (!item) {
      return res.status(404).json({ error: true, message: 'Watchlist item not found or unauthorized.' })
    }

    const updated = await prisma.watchlistItem.update({
      where: { id: itemId },
      data: {
        targetPrice: targetPrice !== undefined ? (targetPrice ? parseFloat(targetPrice) : null) : item.targetPrice,
        alertEnabled: alertEnabled !== undefined ? alertEnabled : item.alertEnabled
      }
    })

    return res.status(200).json({ success: true, item: updated })
  } catch (error) {
    next(error)
  }
}

export async function deleteItem(req: AuthenticatedRequest, res: Response, next: NextFunction): Promise<any> {
  try {
    const userId = req.user!.id
    const { itemId } = req.params

    // Verify ownership
    const item = await prisma.watchlistItem.findFirst({
      where: {
        id: itemId,
        watchlist: { userId }
      }
    })

    if (!item) {
      return res.status(404).json({ error: true, message: 'Watchlist item not found or unauthorized.' })
    }

    await prisma.watchlistItem.delete({ where: { id: itemId } })

    return res.status(200).json({ success: true, message: 'Stock removed from watchlist successfully.' })
  } catch (error) {
    next(error)
  }
}

export async function deleteWatchlist(req: AuthenticatedRequest, res: Response, next: NextFunction): Promise<any> {
  try {
    const userId = req.user!.id
    const { id } = req.params

    const wl = await prisma.watchlist.findFirst({
      where: { id, userId }
    })

    if (!wl) {
      return res.status(404).json({ error: true, message: 'Watchlist not found or unauthorized.' })
    }

    await prisma.watchlist.delete({ where: { id } })

    return res.status(200).json({ success: true, message: 'Watchlist deleted successfully.' })
  } catch (error) {
    next(error)
  }
}

File: fastapi_backend/routers/watchlist.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select
from pydantic import BaseModel
from typing import Optional

from core.database import get_db, Watchlist, WatchlistItem

router = APIRouter(prefix="/api/watchlists", tags=["watchlists"])

# Hardcoded guest user id since auth is removed
GUEST_USER_ID = "guest"

class CreateWatchlistBody(BaseModel):
    name: str

class AddItemBody(BaseModel):
    symbol: str
    targetPrice: Optional[float] = None
    alertEnabled: bool = False

class UpdateItemBody(BaseModel):
    targetPrice: Optional[float] = None
    alertEnabled: Optional[bool] = None

def _wl_dict(wl: Watchlist) -> dict:
    return {
        "id": wl.id, "userId": wl.user_id, "name": wl.name,
        "createdAt": wl.created_at.isoformat(), "updatedAt": wl.updated_at.isoformat(),
        "items": [_item_dict(i) for i in (wl.items or [])]
    }

def _item_dict(item: WatchlistItem) -> dict:
    return {
        "id": item.id, "watchlistId": item.watchlist_id,
        "symbol": item.symbol, "targetPrice": item.target_price,
        "alertEnabled": item.alert_enabled,
        "createdAt": item.created_at.isoformat()
    }

@router.get("")
async def get_watchlists(db: AsyncSession = Depends(get_db)):
    result = await db.execute(
        select(Watchlist).order_by(Watchlist.created_at.desc())
    )
    watchlists = result.scalars().all()
    for wl in watchlists:
        items_result = await db.execute(
            select(WatchlistItem).where(WatchlistItem.watchlist_id == wl.id)
        )
        wl.items = items_result.scalars().all()
    return {"success": True, "watchlists": [_wl_dict(w) for w in watchlists]}

@router.post("", status_code=201)
async def create_watchlist(body: CreateWatchlistBody, db: AsyncSession = Depends(get_db)):
    if not body.name.strip():
        raise HTTPException(400, "Please provide a valid watchlist name.")
    wl = Watchlist(user_id=GUEST_USER_ID, name=body.name.strip())
    wl.items = []
    db.add(wl)
    await db.commit()
    await db.refresh(wl)
    return {"success": True, "watchlist": _wl_dict(wl)}

@router.post("/{watchlist_id}/items", status_code=201)
async def add_item(watchlist_id: str, body: AddItemBody, db: AsyncSession = Depends(get_db)):
    if not body.symbol:
        raise HTTPException(400, "Please specify a company stock symbol.")
    result = await db.execute(select(Watchlist).where(Watchlist.id == watchlist_id))
    wl = result.scalar_one_or_none()
    if not wl:
        raise HTTPException(404, "Watchlist not found.")
    existing = await db.execute(
        select(WatchlistItem).where(
            WatchlistItem.watchlist_id == watchlist_id,
            WatchlistItem.symbol == body.symbol.upper()
        )
    )
    if existing.scalar_one_or_none():
        raise HTTPException(409, "Stock already exists inside this watchlist.")
    item = WatchlistItem(
        watchlist_id=watchlist_id, symbol=body.symbol.upper(),
        target_price=body.targetPrice, alert_enabled=body.alertEnabled
    )
    db.add(item)
    await db.commit()
    await db.refresh(item)
    return {"success": True, "item": _item_dict(item)}

@router.patch("/items/{item_id}")
async def update_item(item_id: str, body: UpdateItemBody, db: AsyncSession = Depends(get_db)):
    result = await db.execute(select(WatchlistItem).where(WatchlistItem.id == item_id))
    item = result.scalar_one_or_none()
    if not item:
        raise HTTPException(404, "Watchlist item not found.")
    if body.targetPrice is not None:
        item.target_price = body.targetPrice
    if body.alertEnabled is not None:
        item.alert_enabled = body.alertEnabled
    await db.commit()
    await db.refresh(item)
    return {"success": True, "item": _item_dict(item)}

@router.delete("/items/{item_id}")
async def delete_item(item_id: str, db: AsyncSession = Depends(get_db)):
    result = await db.execute(select(WatchlistItem).where(WatchlistItem.id == item_id))
    item = result.scalar_one_or_none()
    if not item:
        raise HTTPException(404, "Watchlist item not found.")
    await db.delete(item)
    await db.commit()
    return {"success": True, "message": "Stock removed from watchlist successfully."}

@router.delete("/{watchlist_id}")
async def delete_watchlist(watchlist_id: str, db: AsyncSession = Depends(get_db)):
    result = await db.execute(select(Watchlist).where(Watchlist.id == watchlist_id))
    wl = result.scalar_one_or_none()
    if not wl:
        raise HTTPException(404, "Watchlist not found.")
    await db.delete(wl)
    await db.commit()
    return {"success": True, "message": "Watchlist deleted successfully."}
